contracts/src/v0.8/dev/VRFSubscriptionBalanceMonitor.sol

```diff
@@ -0,0 +1,314 @@
+      }
+      if (subscriptionIds[idx] == 0) {
+        revert InvalidWatchList();
+      }
+      if (topUpAmountsJuels[idx] <= minBalancesJuels[idx]) {
+        revert InvalidWatchList();
+      }
+      s_targets[subscriptionIds[idx]] = Target({
+        isActive: true,
+        minBalanceJuels: minBalancesJuels[idx],
+        topUpAmountJuels: topUpAmountsJuels[idx],
+        lastTopUpTimestamp: 0
+      });
+    }
+    s_watchList = subscriptionIds;
+  }
+
+  /**
+   * @notice Gets a list of subscriptions that are underfunded.
+   * @return list of subscriptions that are underfunded
+   */
+  function getUnderfundedSubscriptions() public view returns (uint64[] memory) {
+    uint64[] memory watchList = s_watchList;
+    uint64[] memory needsFunding = new uint64[](watchList.length);
+    uint256 count = 0;
+    uint256 minWaitPeriod = s_minWaitPeriodSeconds;
+    uint256 contractBalance = LINKTOKEN.balanceOf(address(this));
+    Target memory target;
+    for (uint256 idx = 0; idx < watchList.length; idx++) {
+      target = s_targets[watchList[idx]];
+      (uint96 subscriptionBalance, , , ) = COORDINATOR.getSubscription(watchList[idx]);
+      if (
+        target.lastTopUpTimestamp + minWaitPeriod <= block.timestamp &&
+        contractBalance >= target.topUpAmountJuels &&
+        subscriptionBalance < target.minBalanceJuels
+      ) {
+        needsFunding[count] = watchList[idx];
+        count++;
+        contractBalance -= target.topUpAmountJuels;
+      }
+    }
+    if (count != watchList.length) {
+      assembly {
+        mstore(needsFunding, count)
+      }
+    }
+    return needsFunding;
+  }
+
+  /**
+   * @notice Send funds to the subscriptions provided.
+   * @param needsFunding the list of subscriptions to fund
+   */
+  function topUp(uint64[] memory needsFunding) public whenNotPaused {
+    uint256 minWaitPeriodSeconds = s_minWaitPeriodSeconds;
+    uint256 contractBalance = LINKTOKEN.balanceOf(address(this));
+    Target memory target;
+    for (uint256 idx = 0; idx < needsFunding.length; idx++) {
+      target = s_targets[needsFunding[idx]];
+      (uint96 subscriptionBalance, , , ) = COORDINATOR.getSubscription(needsFunding[idx]);
+      if (
+        target.isActive &&
+        target.lastTopUpTimestamp + minWaitPeriodSeconds <= block.timestamp &&
+        subscriptionBalance < target.minBalanceJuels &&
+        contractBalance >= target.topUpAmountJuels
+      ) {
+        bool success = LINKTOKEN.transferAndCall(
+          address(COORDINATOR),
+          target.topUpAmountJuels,
+          abi.encode(needsFunding[idx])
+        );
+        if (success) {
+          s_targets[needsFunding[idx]].lastTopUpTimestamp = uint56(block.timestamp);
+          contractBalance -= target.topUpAmountJuels;
+          emit TopUpSucceeded(needsFunding[idx]);
+        } else {
+          emit TopUpFailed(needsFunding[idx]);
+        }
+      }
+      if (gasleft() < MIN_GAS_FOR_TRANSFER) {
```

Comment:
I feel like we should emit something here - in order to increase the gas allowance for this upkeep.